### ISSUE TITLE
My (public) ElectrumX server became too popular, and now it often throttles and doesn't let me connect!

### ISSUE THREAD

**Issue by maltokyo** [spesmilo/electrumx#129]
I have 48GB RAM, and lots of it allocated to ElectrumX (as per the recommended max in the docs), so resources are not an issue.  I also have 1000 sessions limit.  Often 700 users or more connected.  This seems fine as far as server load is concerned, but the throttling begins frequently anyway.

How can I either:
- solve this, so no throttling happens at all (unless the server runs out of RAM)
- or, alternatively, at least prioritize my OWN connection to the server, it is super frustrating getting throttled off my own server (haha..)

Please give any advice you might have, otherwise, I will just make it all private and only for me, but I would like to avoid that, as long as I have spare server capacity to burn for the good of the network.

**Comment by SomberNight** [comment]
> fine as far as server load is concerned, but the throttling begins frequently anyway.

What kind of throttling do you mean/experience?

**Comment by maltokyo** [comment]
the log says something like "10 concurrents, throttling to 0", or something to this meaning, and when it starts doing that I cannot connect at all, or sporadically.  Now it is slower, so I can connect and cant see these log entries, when I find it again I will paste it here.

**Comment by SomberNight** [comment]
How does your client connect to your server? What incoming IP does the server see for the client?

**Comment by maltokyo** [comment]
The client is the standard Electrum wallet.  The server log sees/shows my public IP address (the server is in a different location).  I use SSL port 50002 only (the server does not listen on 50001 or other ports).

**Comment by maltokyo** [comment]
So it is happening now, cannot connect at all. Even after a restart which is strange. 

Log attached as a screenshot. Sorry only have mobile. 

Kindly let me know what I could get to resolve this. Ill soon get so frustrated that I'll just make it private over tor. 

Thank you!!

[IMG_20210503_025720.jpg](https://user-images.githubusercontent.com/1388507/116833813-8eec0a80-abbb-11eb-8f51-b43b7e47b473.jpg)

**Comment by SomberNight** [comment]
Could you test with disabled resource usage limits?
```
COST_SOFT_LIMIT=0
COST_HARD_LIMIT=0
```

(see https://electrumx-spesmilo.readthedocs.io/en/latest/environment.html#resource-usage-limits for more info)

**Comment by maltokyo** [comment]
Thank you, trying now

**Comment by maltokyo** [comment]
Runs very smoothly now, even with 1000 connections.  Looks good.

**Comment by maltokyo** [comment]
Everything is very stable now.  Between 500-1000 connections being served without any hitch, for days now.
Perhaps for servers with a lot of resource, a setting like the above would be helpful to be more prominent in the docs (in terns of exactly what it is doing)

**Comment by SomberNight** [comment]
Well it already has its own section in the docs; that's what I linked to. Or do you mean specifically this `COST_SOFT_LIMIT=0; COST_HARD_LIMIT=0` should be mentioned?

Btw I think maybe what might have been the culprit is the default for `SESSION_GROUP_BY_SUBNET_IPV4`. It defaults to /24; maybe it should be changed to /32. This could explain why your IP got throttled.

**Comment by maltokyo** [comment]
> Well it already has its own section in the docs; that's what I linked to. Or do you mean specifically this `COST_SOFT_LIMIT=0; COST_HARD_LIMIT=0` should be mentioned?

Yes - was mentioning to add a comment something like "set these to 0 in case you see 'disconnected whilst throttled' message in the log when you think that you have enough server power and it should not be throttled under any conditions".  
At least this way, a solution would be available to people using Google to search for the error text.  Just my 2 sats.


> Btw I think maybe what might have been the culprit is the default for `SESSION_GROUP_BY_SUBNET_IPV4`. It defaults to /24; maybe it should be changed to /32. This could explain why your IP got throttled.

That could be true too - because it is running in docker, on its own IP subnet, so the requests from the internet or the data center network might be treated differently.
